### PR TITLE
GAME TREE SCENARIOS

### PR DESCRIPTION
**Pull request by sriharikapu**
Covers various scenarios of game tress
Implemented the following codes 

Actions included

- [x] buyer-commit
- [x] buyer-complain
- [x] buyer-redeem
- [x] buyer-refund
- [x] seller-cancel
- [x] seller-commit
- [x] seller-admitFault

Helpers included 

- [x] Delay
- [x] Formatter
- [x]  Check Balances

Also included

- [x] sign Transactions

Tests

- [x] Modularized

Scenarios Covered

- [x] SELLER LISTS A VOUCHER FOR SALE
- [x] BUYER PURCHASES A VOUCHER
- [x] BUYER REDEEMS A COMMITTED VOUCHER
- [x] BUYER CLAIMS REFUND FOR A COMMITTED VOUCHER
- [x] BUYER COMPLAINS ABOUT REDEEMED VOUCHER
- [x] BUYER COMPLAINS ABOUT REFUND VOUCHER
- [x] SELLER ACCEPTS REDEEM FAULT ON COMPLAIN
- [x] SELLER ACCEPTS REFUND FAULT ON COMPLAIN

[Read More](https://github.com/bosonprotocol/contracts/blob/gametree-trace/test/readme.md)

### CODE REVIEW

**Review by hswopeams — CHANGES_REQUESTED** [review]
A couple of things

- It looks like the formatter didn't like something. That's what made the pipeline fail. Try running `./go tests:format_fix` and then recommitting
- Please move the actions folder and everything in it into a folder called testGameTree or similar. Right now it's not clear that the actions folder is for tests
- The tests output using console.log, but I don't see any assertions. It's assertions that make a test a test. For instance when balances are checked in "Check Balances" blocks, there should be an assertion instead of a `console.log`

File: testGameTree/5-buyer-commit-refund.js
Comment (HristiyanG) [inline testGameTree/5-buyer-commit-refund.js:14]: redundant variable

**Review by HristiyanG — CHANGES_REQUESTED** [review]

**Comment by HristiyanG** [comment]
Just finished my review, left some comments there, but just still some general questions: 
- Why the old test were deleted? There were roughly about 400 test. Do these scenarios cover each and every missing test?
	- if that is not the case, please bring back the old test
	- please consider writing a script, responsible for the user interactions, and leave the `test` script for the unit testing
- Agreed with Heather that the `actions` folder needs to be moved into the tests, maybe renamed to testGameTreeActions or similar
- Agreed with Heather that the tests should be using assertions
- Do we need to test those actions over a testnet at all?
	- we can do it by running 2 different scripts then, 1 for testnet (e.g. Rinkeby), 1 locally
	- this way while run locally, we should not bother to fill in the config file. We can get the accounts from the local rpc, deploy a contract and fill in the required params automatically
- There should be something at the end indicating that the tests are completed. Currently when the process finishes I was misled that I am doing something wrong as there is not anything which prompts me how tests went
- Maybe a little more detailed README file would be good. I don't think the delay should be that long (30000) if we test the flow locally. Maybe even we could move it into the configs?
- `constant.js` exported from the `test` folder, and now moved to actions/helpers/constants. Do we need to do that. IMHO this should be in the test folder as it's own purpose is to serve the tests
- left some comments for removing the empty lines, commented lines and other

File: testGameTree/1-seller-create.js
Comment (sriharikapu) [reply testGameTree/1-seller-create.js]: removed it

File: testGameTree/buyer/refundVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/refundVoucher.js]: removed them

File: testGameTree/buyer/refundVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/refundVoucher.js]: removed

File: testGameTree/helpers/checkBalance.js
Comment (sriharikapu) [reply testGameTree/helpers/checkBalance.js:27]: resolved

File: testGameTree/helpers/formatter.js
Comment (sriharikapu) [reply testGameTree/helpers/formatter.js:5]: removed the commented line

File: testGameTree/helpers/formatter.js
Comment (sriharikapu) [reply testGameTree/helpers/formatter.js:19]: removed the commented lines

File: testGameTree/seller/cancelVoucherSet.js
Comment (sriharikapu) [reply testGameTree/seller/cancelVoucherSet.js]: removed

File: testGameTree/seller/cancelVoucherSet.js
Comment (sriharikapu) [reply testGameTree/seller/cancelVoucherSet.js:20]: removed

File: testGameTree/seller/cancelVoucherSet.js
Comment (sriharikapu) [reply testGameTree/seller/cancelVoucherSet.js]: removed

File: testGameTree/seller/cancelVoucherSet.js
Comment (sriharikapu) [reply testGameTree/seller/cancelVoucherSet.js]: removed

File: testGameTree/seller/cancelVoucherSet.js
Comment (sriharikapu) [reply testGameTree/seller/cancelVoucherSet.js]: removed

File: testGameTree/seller/createVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/createVoucher.js]: removed

File: testGameTree/seller/createVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/createVoucher.js]: removed

File: testGameTree/seller/createVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/createVoucher.js]: removed

File: testGameTree/seller/createVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/createVoucher.js]: removed

File: testGameTree/seller/createVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/createVoucher.js]: removed

File: testGameTree/seller/createVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/createVoucher.js]: removed

File: testGameTree/seller/faultVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/faultVoucher.js]: removed

File: testGameTree/seller/faultVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/faultVoucher.js]: removed

File: testGameTree/seller/faultVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/faultVoucher.js]: removed

File: testGameTree/seller/faultVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/faultVoucher.js]: removed

File: testGameTree/seller/faultVoucher.js
Comment (sriharikapu) [reply testGameTree/seller/faultVoucher.js]: removed

File: testGameTree/2-seller-create-cancel.js
Comment (sriharikapu) [reply testGameTree/2-seller-create-cancel.js:8]: removed

File: testGameTree/4-buyer-commit-redeem.js
Comment (sriharikapu) [reply testGameTree/4-buyer-commit-redeem.js]: removed

File: testGameTree/6-buyer-commit-redeem-complain.js
Comment (sriharikapu) [reply testGameTree/6-buyer-commit-redeem-complain.js]: removed

File: testGameTree/7-buyer-commit-refund-complain.js
Comment (sriharikapu) [reply testGameTree/7-buyer-commit-refund-complain.js]: removed

File: testGameTree/8-buyer-commit-redeem-complain-seller-fault.js
Comment (sriharikapu) [reply testGameTree/8-buyer-commit-redeem-complain-seller-fault.js:17]: removed

File: testGameTree/9-buyer-commit-refund-complain-seller-fault.js
Comment (sriharikapu) [reply testGameTree/9-buyer-commit-refund-complain-seller-fault.js:13]: removed

File: testGameTree/9-buyer-commit-seller-fault.js
Comment (sriharikapu) [reply testGameTree/9-buyer-commit-seller-fault.js]: removed

File: testGameTree/9-buyer-commit-seller-fault.js
Comment (sriharikapu) [reply testGameTree/9-buyer-commit-seller-fault.js]: removed

File: testGameTree/9-buyer-commit-seller-fault.js
Comment (sriharikapu) [reply testGameTree/9-buyer-commit-seller-fault.js]: removed

File: testGameTree/buyer/commitVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/commitVoucher.js:45]: removed

File: testGameTree/buyer/commitVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/commitVoucher.js:65]: mofied

File: testGameTree/buyer/compainVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/compainVoucher.js:19]: removed

File: testGameTree/buyer/compainVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/compainVoucher.js]: removed

File: testGameTree/buyer/compainVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/compainVoucher.js]: removed

File: testGameTree/buyer/compainVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/compainVoucher.js]: removed

File: testGameTree/buyer/redeemVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/redeemVoucher.js]: removed

File: testGameTree/buyer/redeemVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/redeemVoucher.js]: removed

File: testGameTree/buyer/redeemVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/redeemVoucher.js]: removed

File: testGameTree/buyer/redeemVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/redeemVoucher.js:19]: removed

File: testGameTree/buyer/redeemVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/redeemVoucher.js]: removed

File: testGameTree/buyer/redeemVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/redeemVoucher.js]: removed

File: testGameTree/buyer/redeemVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/redeemVoucher.js]: removed

File: testGameTree/buyer/refundVoucher.js
Comment (sriharikapu) [reply testGameTree/buyer/refundVoucher.js]: removed

**Comment by sriharikapu** [comment]
Why the old test were deleted? There were roughly about 400 tests. Do these scenarios cover each and every missing test?
 - [x] Brought Back the previous test
 
 Agreed with Heather that the actions folder needs to be moved into the tests, maybe renamed to testGameTreeActions or similar
 - [x] Renamed to testGameTree
 
 Agreed with Heather that the tests should be using assertions
 - [x] Added More assertions to tests
 
Do we need to test those actions over a testnet at all?
 - [x] You can set the provider to local or any testnet url so you can run them
 
There should be something at the end indicating that the tests are completed. Currently when the process finishes I was misled that I am doing something wrong as there is not anything which prompts me how tests went
- [x] once the tests finish you will be able to see the number of tests passing, 
 
Maybe a little more detailed README file would be good. I don't think the delay should be that long (30000) if we test the flow locally. Maybe even we could move it into the configs?
- [x] delay was set to 30000ms initially because of the delay in transactions getting confirmed lately, I had kept the delay to the config so you can customized for local test and testnet etc

constant.js exported from the test folder, and now moved to actions/helpers/constants. Do we need to do that. IMHO this should be in the test folder as it's own purpose is to serve the tests
- [x] Actually we don't need all the variables in the constants from the `constants.js`, I had removed the unused variables. also we can link to the previous constants

left some comments for removing the empty lines, commented lines and other
- [x] removed unnecessary empty lines mentioned

pipeline should succeed before merging
- [x]  made sure that pipeline passes

**Review by hswopeams — APPROVED** [review]

**Comment by sriharikapu** [comment]
@hswopeams I think we can merge this once @HristiyanG approves it. Also I checked at my end everything works as expected.

**Comment by hswopeams** [comment]
Changed gametree tests to run locally on Ganace instance started by the (Ruby) `go` script. Discussed with Holly, and it was decided that running locally is enough. There's no need to add complexity by adding config to also run against a testnet.

**Comment by thecryptofruit** [comment]
@damir-bajramovic , please have a look if this is in your card, otherwise close. thanks

**Comment by mischat** [comment]
All of the Game Tree has been incorporated into our unit test suite, we are happy that we are covering all of the game tree and more in `main` of this repo. This branch will be deleted.

**Pull request closed without merging**
